meta_error.go

```diff
@@ -0,0 +1,112 @@
+package microerror
+
+import (
+	"encoding/json"
+
+	"github.com/juju/errgo"
+)
+
+type metaError struct {
+	cause    error
+	metadata map[string]string
+}
+
+// NewMetaError creates a new instance of metaError. Its purpose is to act as
+// container for meta information associated to a specific error. Package
+// internally metaError instances can be used independent from each other. The
+// specific error type matching can be used as usual. Using the meta error looks
+// as follows. In the beginning is a usual error defined on its own. This error
+// is the root cause once emitted during runtime.
+//
+//     var notEnoughWorkersError = microerror.New("not enough workers")
+//
+// Emitting the error defined above looks as follows. In the described situation
+// something happens to cause the error propagation. The defined error is given
+// to NewMetaError in order to annotate it with the given context information.
+// The meta error is masked as usual.
+//
+//     return microerror.Mask(microerror.NewMetaError(notEnoughWorkersError, map[string]string{
+//         "available",   strconv.Itoa(info.Workers.CountPerCluster.Max),
+//         "description", "The amount of requested guest cluster workers exceeds the available number of host cluster nodes.",
+//         "kind",        "notEnoughWorkersError",
+//         "ops-recipe",  "https://github.com/giantswarm/ops-recipes/blob/master/349-not-enough-workers.md",
+//         "requested",   strconv.Itoa(len(request.Cluster.Workers)),
+//     }))
+//
+// Down the stack contextual information can be retrieved from the masked error.
+// In the described example the key kind results in the dispatched value
+// notEnoughWorkersError.
+//
+//     val, err := microerror.FromMetaError(err, "kind")
+//     if err != nil {
+//         return microerror.Mask(err)
+//     }
+//
+func NewMetaError(err error, meta map[string]string) error {
```

Comment:
I really don't like this takes a map. I would strongly prefer to have a specific type here. I don't see randomly extracted key value pairs being useful.

Also I don't think it even has to be a constructor. We can just have an error type which implements error interface.

Reply:
- Thanks for the feedback, but you need to give me a bit more. Please provide alternatives. Explain what you think it should look like. Some pseudo code is good enough. I do not know how to improve my proposal based on your comments here. 
- I do not understand the problem with the map. What specific type do you have in mind? I do not follow here. I think it is good to have the flexibility of the map so we can dispatch arbitrary key-value pairs. Why this is useful is just shown in the example in the huge comment above the constructor. 
- I do not know how this can work without a constructor. There is a specific UX design here, which does not work when there is a specific error type. The example above explains this quite well.

Reply:
I think quite opposite. Too much flexibility becomes hard to maintain, enforce and scary to use. With a well defined type you handover that job (at least partially) to the compiler.

By a specific error type I meant something like:

```
type Error struct {
	Description string
	Kind        string
	OpsRecipe   *url.URL
}

func (e Error) Error() string {
	...
}

func (m Error) MarshalJSON() ([]byte, error) {
	...
}
```

Then it can be just initialized like that:

```
Error{
	Description: "The amount of requested guest cluster workers exceeds the available number of host cluster nodes.",
	Kind:        "notEnoughWorkersError",
	OpsRecipe:   "https://github.com/giantswarm/ops-recipes/blob/master/349-not-enough-workers.md",
}
```

If we see the need of extending it we can just add another field when it's clarified.

Extracting information can also become clearer and safer. We can provide `ToError(err error) (Error, ok)` which basically finds if `err` cause if of `Error` type.

Reply:
Thinking more about that I would even think about making `Kind` a strong type and provide consts here.

Reply:
I like this, it is just that it does not integrate in how we do error handling now so we would either have to rewrite all the errors or we lose the ability to match against error types using our error matchers within the same process. Again, check the description above the constructor. There is a story being told. Tell me your story. I do not see how this integrates into the current error UX design and how we do not lose anything we have. The proposal I made is composable and integrates seemlessly in what we have. Now, I am fine when we want to have a predefined structure for the contextual information. But I think the rest still stays the same then. Which again contradicts a bit with how this works out in the wild.

Reply:
This is what I mean. We would have to create these errors from scratch and rewrite basically everything. That is pretty odd. I would like to have an approach which integrates seamlessly in a composing way without the necessity of rewriting all errors.

Reply:
So I think what I proposed is a viable solution with respect to these aspects. As I said, I like having the map replaced with a predefined structure.

Reply:
I know, but you cannot just annotate `invalidConfigError`. You have to replace it with the new type. That is my point. This is rather a disadvantage.

Reply:
When you want to annotate it you already doing changes. When you have a strong type you do it in a single place - you change the error variable - thanks to `Maskf` being backward compatible. I'd say doing changes in every single `return` statement is disadvantage.


```
return microerror.Mask(microerror.NewMetaError(invalidConfig, map[string]string{
    "description", fmt.Sprintf("%T.Name must not be empty", config),
    "kind",        "invalidConfig",
    "ops-recipe",  "https://github.com/giantswarm/ops-recipes/blob/master/349-invalid-config.md",
}))

return microerror.Mask(microerror.NewMetaError(invalidConfig, map[string]string{
    "description", fmt.Sprintf("%T.OtherName must not be empty", config),
    "kind",        "invalidConfig",
    "ops-recipe",  "https://github.com/giantswarm/ops-recipes/blob/master/349-invalid-config.md",
}))
```

Instead of:

```
var invalidConfigError = microerror.Error{
	Kind:        "invalidConfig",
	OpsRecipe:   "https://github.com/giantswarm/ops-recipes/blob/master/349-invalid-config.md",
}
```

And then don't touch return statements. Profit.